---
title: "Simulations with GTEx genotypes and effect structure similar to GTEx gene expression"
author: "Fabio Morgante"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r set opts, warning=FALSE, message=FALSE}
###Set options
options(stringsAsFactors=FALSE)

###Load libraries
library(dscrutils)
library(ggplot2)
library(cowplot)
library(scales)

###Function to convert dscquery output from list to data.frame suitable for plotting
convert_dsc_to_dataframe_gtex <- function(dsc){
  ###Data.frame to store the results after convertion
  dsc_df <- data.frame()
  
  ###Get length of list elements 
  n_elem <- length(dsc$DSC)
  
  ###Loop through the dsc list
  for(i in 1:n_elem){
    ##Prepare vectors making up the final data frame
    r_scalar <- dsc$simulate.r[i]
    p_causal <- rep(dsc$simulate.p_causal[i], times=r_scalar)
    r <- rep(dsc$simulate.r[i], times=r_scalar)
    response <- 1:r_scalar
    pve <- rep(dsc$simulate.pve[i], times=r_scalar)
    simulate <- rep(dsc$simulate[i], times=r_scalar)
    fit <- rep(dsc$fit[i], times=r_scalar)
    score <- rep(dsc$score[i], times=r_scalar)
    score.err <- dsc$score.err[[i]]
    timing <- rep(dsc$fit.time[i], times=r_scalar)
    
    ##Build the data frame
    df <- data.frame(p_num_caus=p_causal, r=r, response=response, pve=pve,  
                     scenario=simulate, method=fit, score_metric=score, score_value=score.err, time=timing)
    dsc_df <- rbind(dsc_df, df)
  }
  
  ###Get number of genes
  n_methods <- length(unique(dsc$fit))
  n_metrics <- length(unique(dsc$score))
  n_genes <- n_elem/n_methods/n_metrics
 
  ###Compute replicates
  reptot <- rep(rep(1:n_genes, each=dsc$simulate.r[1]), n_methods*n_metrics)
  
  dsc_df <- data.frame(rep=reptot, dsc_df)
  
  return(dsc_df)
}

###Function to compute rmse (relative to mr_mash_em_can)
compute_rrmse <- function(dsc_plot, log10_scale=FALSE){
  dsc_plot <- transform(dsc_plot, experiment=paste(rep, response, scenario, sep="-"))
  t <- 0
  for (i in unique(dsc_plot$experiment)) {
    t <- t+1
    rmse_data  <- dsc_plot[which(dsc_plot$experiment == i & dsc_plot$score_metric=="scaled_rmse"), ]
    mse_mr_mash_consec_em <- rmse_data[which(rmse_data$method=="mr_mash_em_can"), "score_value"]
    if(!log10_scale)
      rmse_data$score_value <- rmse_data$score_value/mse_mr_mash_consec_em
    else
      rmse_data$score_value <- log10(rmse_data$score_value/mse_mr_mash_consec_em)
    rmse_data$score_metric <- "scaled_rrmse"
    if(t>1){
      rmse_data_tot <- rbind(rmse_data_tot, rmse_data)
    } else if(t==1){
      rmse_data_tot <- rmse_data
    }
  }
  
  rmse_data_tot$experiment <- NULL
  
  return(rmse_data_tot)
}

###Set some quantities used in the following plots
colors <- c("cyan", "skyblue", "dodgerblue", "mediumblue", "limegreen", "green", "gold", "orange", "red", "firebrick", "darkmagenta", "mediumpurple")
```

# PVE=0.2

```{r load data, warning=FALSE, message=FALSE}
###Load the dsc results
dsc_out <- dscquery("output/mvreg_all_genes_prior_GTExrealX_indepV_sharedB_2blocksr10", 
                    c("simulate.p_causal", "simulate.r", "simulate.r_causal", "simulate.pve", "simulate.B_cor", "simulate.w", 
                      "simulate.B_scale", "simulate.V_cor", "simulate", "fit", "score", "score.err", "fit.time"), 
                    groups="fit: mr_mash_em_can, mr_mash_em_data, mr_mash_em_dataAndcan, mr_mash_em_dataAndcan_dropcomp, mlasso, mtlasso, enet",
                    verbose=FALSE)

###Obtain simulation parameters
prop_testset <- 0.2
p_causal <- unique(dsc_out$simulate.p_causal)
r <- unique(dsc_out$simulate.r)
r_causal <- eval(parse(text=unique(dsc_out$simulate.r_causal)))
pve <- unique(dsc_out$simulate.pve)
w <- eval(parse(text=unique(dsc_out$simulate.w)))
B_cor <- eval(parse(text=unique(dsc_out$simulate.B_cor)))
B_scale <- eval(parse(text=unique(dsc_out$simulate.B_scale)))
V_cor <- unique(dsc_out$simulate.V_cor)

###Remove list elements that are not useful anymore
dsc_out$simulate.r_causal <- NULL
dsc_out$simulate.w <- NULL
dsc_out$simulate.B_cor <- NULL
dsc_out$simulate.B_scale <- NULL
dsc_out$simulate.V_cor <- NULL
```


## Simulation set up

The results below are based on simulations with GTEx v8 cis-genotypes for 838 samples, with `r p_causal` causal variants, `r r` responses with a per-response proportion of variance explained (PVE) of `r pve`. Causal effects, B, were drawn from $w_1 MVN(0, Sigma1) + w_2 MVN(0, Sigma2)$, where $w_1$ = `r w[1]` and $w_2$ = `r w[2]` Sigma1 is such that it achieves a correlation between responses of `r B_cor[1]` and a scale of `r B_scale[1]` and Sigma2 is such that it achieves a correlation between responses of `r B_cor[2]` and a scale of `r B_scale[2]`. The first component of the mixture applies to responses `r r_causal[[1]]` while the second component applies to responses `r r_causal[[2]]`. This structure is meant to mirror that of brain tissues and non-brain tissues in GTEx. The responses, Y, were drawn from MN(XB, I, V), where V is such that it achieves a correlation between responses of `r V_cor` and a scale defined by PVE, in the case of cuasal responses. Non causal responses are from N(0, 1).

2000 genes were simulated and univariate summary statistics were obtained by simple linear regression in the training data (`r (1-prop_testset)*100`% of the data. The indexes of the training-test individuals were the same for all the datasets). These regression coefficients and standard errors were used as input in the *mash* pipeline (from Gao) to compute data-driven covariance matrices (up to the ED step included). In particular, the top variable per gene was used to define a "strong" set and 4 random variables per gene were used to define a "random" set. Covariance matrices were estimated using *flash*, PCA (including the top 3 PCs), and the empirical covariance matrix.  

The first 50 genes were used for the prediction analysis. *mr.mash* was fitted to the training data, updating V (imposing a diagonal structure) and updating the prior weights using EM updates. The mixture prior consisted of components defined by:

* canonical matrices corresponding to different settings of effect sharing/specificity (i.e., singletons, independent, low heterogeneity, medium heterogeneity, high heterogeneity, shared) plus the spike. 

* data-driven matrices estimated as described above plus the spike.

* both canonical and data-driven matrices plus the spike.

* both canonical and data-driven matrices plus the spike, dropping components with $w_0 < 1e-8$.

The covariances matrices were scaled by a grid of values computed from the univariate summary statistics as in the *mash* paper. The posterior mean of the regression coefficients were initialized to the estimates of the group-LASSO. The mixture weights were initialized with the proportion of zero-coefficients from the group-LASSO estimate as the weight on the spike and the proportion of non-zero-coefficients split equally among the remaining components.Convergence was declared when the maximum difference in the ELBO between two successive iterations was smaller than 1e-2. 

Then, responses were predicted on the test data (`r prop_testset*100`% of the data). 

Here, we evaluate the accuracy of prediction assessed by $r^2$ and bias (slope) from the regression of the true response on the predicted response, and the relative root mean square error (rRMSE) scaled by the standard deviation of the true responses in the test data. The boxplots are across the 50 genes.

## mr.mash vs other methods

Here, we compare *mr.mash* to the multivariate versions of LASSO as implemented in *glmnet*. The form of the penalty id the following: $\lambda[(1-\alpha)/2 ||\mathbf{\beta}_j||^2_2 + \alpha ||\mathbf{\beta}_j||_2]$. $\lambda$ is chosen by cross-validation in the training set. We also compare to the [*UTMOST* implementation](https://users.rcc.uchicago.edu/~aksarkar/gtex-pred/lasso.html) by Abhishek, and the response-by-response elastic net. All the methods were using 1 thread.

```{r accuracy vs others, fig.height=12, fig.width=15, message=FALSE}
###Convert from list to data.frame for plotting
dsc_plots <- convert_dsc_to_dataframe_gtex(dsc_out)

###Compute rmse score (relative to mr_mash_consec_em) and add it to the data
rrmse_dat <- compute_rrmse(dsc_plots)
dsc_plots <- rbind(dsc_plots, rrmse_dat)

###Remove mse from scores and keep only methods wanted
dsc_plots <- dsc_plots[which(dsc_plots$score_metric!="scaled_rmse" ), ]

###Create factor version of method
dsc_plots$method_fac <- factor(dsc_plots$method, levels=c("mr_mash_em_can", "mr_mash_em_data",
                                                          "mr_mash_em_dataAndcan", "mr_mash_em_dataAndcan_dropcomp", 
                                                          "mlasso", "mtlasso", "enet"),
                                                labels=c("mr_mash_can", "mr_mash_data", "mr_mash_both", 
                                                         "mr_mash_both_drop", "mlasso", "mtlasso", "enet"))

###Create factor version of response
dsc_plots$response_fac <- as.factor(dsc_plots$response)

###Create plots
p_rrmse <- ggplot(dsc_plots[which(dsc_plots$score_metric=="scaled_rrmse"), ], aes_string(x = "response_fac", y = "score_value", fill = "method_fac")) +
  geom_boxplot(color = "black", outlier.size = 1, width = 0.85) +
  scale_fill_manual(values = colors) +
  labs(x = "Response", y = "Error", title = "rRMSE", fill="Method") +
  geom_hline(yintercept=1, linetype="dotted", size=1) +
  theme_cowplot(font_size = 20) +
  theme(plot.title = element_text(hjust = 0.5))

print(p_rrmse)
```

Let's now remove outliers from the plots to make things a little clearer.

```{r accuracy no outliers, fig.height=12, fig.width=15, message=FALSE}
p_rrmse_noout <- ggplot(dsc_plots[which(dsc_plots$score_metric=="scaled_rrmse"), ], aes_string(x = "response_fac", y = "score_value", fill = "method_fac")) +
  Ipaper::geom_boxplot2(color = "black",width.errorbar = 0, width = 0.85) +
  scale_fill_manual(values = colors) +
  labs(x = "Response", y = "Error", title = "rRMSE", fill="Method") +
  geom_hline(yintercept=1, linetype="dotted", size=1) +
  theme_cowplot(font_size = 20) +
  theme(plot.title = element_text(hjust = 0.5))

print(p_rrmse_noout)
```

Here, we look at the elapsed time ($log_2$ seconds) of each method. Note that the *mr.mash* run time does not include the run time of group-LASSO (but should be considered since we used it to initialize *mr.mash*).

```{r run time, fig.height=12, fig.width=15, message=FALSE}
dsc_plots_time <- dsc_plots[which(dsc_plots$response==1 & dsc_plots$score_metric=="scaled_rrmse"), 
                          -which(colnames(dsc_plots) %in% c("score_metric", "score_value", "response"))]

p_time <- ggplot(dsc_plots_time, aes_string(x = "method_fac", y = "time", fill = "method_fac")) +
  geom_boxplot(color = "black", outlier.size = 1, width = 0.85) +
  scale_fill_manual(values = colors) +
  scale_y_continuous(trans="log2", breaks = trans_breaks("log2", function(x) 2^x),
                     labels = trans_format("log2", math_format(2^.x))) +
  labs(x = "", y = "Elapsed time (seconds) in log2 scale", title = "Run time", fill="Method") +
  theme_cowplot(font_size = 20) +
  theme(axis.line.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
        plot.title = element_text(hjust = 0.5))

print(p_time)
```


<!-- # PVE=0.05 -->

<!-- ```{r load data low pve, warning=FALSE, message=FALSE} -->
<!-- ###Load the dsc results -->
<!-- dsc_out <- dscquery("output/mvreg_all_genes_prior_GTExrealX_indepV_sharedB_2blocksr10_lowPVE",  -->
<!--                     c("simulate.p_causal", "simulate.r", "simulate.r_causal", "simulate.pve", "simulate.B_cor", "simulate.w",  -->
<!--                       "simulate.B_scale", "simulate.V_cor", "simulate", "fit", "score", "score.err", "fit.time"),  -->
<!--                     groups="fit: mr_mash_em_can, mr_mash_em_data, mr_mash_em_dataAndcan, mr_mash_em_dataAndcan_dropcomp, mlasso, mtlasso, enet", -->
<!--                     verbose=FALSE) -->

<!-- ###Obtain simulation parameters -->
<!-- prop_testset <- 0.2 -->
<!-- p_causal <- unique(dsc_out$simulate.p_causal) -->
<!-- r <- unique(dsc_out$simulate.r) -->
<!-- r_causal <- eval(parse(text=unique(dsc_out$simulate.r_causal))) -->
<!-- pve <- unique(dsc_out$simulate.pve) -->
<!-- w <- eval(parse(text=unique(dsc_out$simulate.w))) -->
<!-- B_cor <- eval(parse(text=unique(dsc_out$simulate.B_cor))) -->
<!-- B_scale <- eval(parse(text=unique(dsc_out$simulate.B_scale))) -->
<!-- V_cor <- unique(dsc_out$simulate.V_cor) -->

<!-- ###Remove list elements that are not useful anymore -->
<!-- dsc_out$simulate.r_causal <- NULL -->
<!-- dsc_out$simulate.w <- NULL -->
<!-- dsc_out$simulate.B_cor <- NULL -->
<!-- dsc_out$simulate.B_scale <- NULL -->
<!-- dsc_out$simulate.V_cor <- NULL -->
<!-- ``` -->

<!-- In this section the set up is exactly the same but with PVE=`r pve`. -->

<!-- ## mr.mash vs other methods -->

<!-- ```{r accuracy vs others low PVE, fig.height=12, fig.width=15, message=FALSE} -->
<!-- ###Convert from list to data.frame for plotting -->
<!-- dsc_plots <- convert_dsc_to_dataframe_gtex(dsc_out) -->

<!-- ###Compute rmse score (relative to mr_mash_consec_em) and add it to the data -->
<!-- rrmse_dat <- compute_rrmse(dsc_plots) -->
<!-- dsc_plots <- rbind(dsc_plots, rrmse_dat) -->

<!-- ###Remove mse from scores and keep only methods wanted -->
<!-- dsc_plots <- dsc_plots[which(dsc_plots$score_metric!="scaled_rmse" ), ] -->

<!-- ###Create factor version of method -->
<!-- dsc_plots$method_fac <- factor(dsc_plots$method, levels=c("mr_mash_em_can", "mr_mash_em_data", -->
<!--                                                           "mr_mash_em_dataAndcan", "mr_mash_em_dataAndcan_dropcomp",  -->
<!--                                                           "mlasso", "mtlasso", "enet"), -->
<!--                                                 labels=c("mr_mash_can", "mr_mash_data", "mr_mash_both",  -->
<!--                                                          "mr_mash_both_drop", "mlasso", "mtlasso", "enet")) -->

<!-- ###Create factor version of response -->
<!-- dsc_plots$response_fac <- as.factor(dsc_plots$response) -->

<!-- ###Create plots -->
<!-- p_rrmse <- ggplot(dsc_plots[which(dsc_plots$score_metric=="scaled_rrmse"), ], aes_string(x = "response_fac", y = "score_value", fill = "method_fac")) + -->
<!--   geom_boxplot(color = "black", outlier.size = 1, width = 0.85) + -->
<!--   scale_fill_manual(values = colors) + -->
<!--   labs(x = "Response", y = "Error", title = "rRMSE", fill="Method") + -->
<!--   geom_hline(yintercept=1, linetype="dotted", size=1) + -->
<!--   theme_cowplot(font_size = 20) + -->
<!--   theme(plot.title = element_text(hjust = 0.5)) -->

<!-- print(p_rrmse) -->
<!-- ``` -->

<!-- Let's now remove outliers from the plots to make things a little clearer. -->

<!-- ```{r accuracy no outliers low PVE, fig.height=12, fig.width=15, message=FALSE} -->
<!-- p_rrmse_noout <- ggplot(dsc_plots[which(dsc_plots$score_metric=="scaled_rrmse"), ], aes_string(x = "response_fac", y = "score_value", fill = "method_fac")) + -->
<!--   Ipaper::geom_boxplot2(color = "black",width.errorbar = 0, width = 0.85) + -->
<!--   scale_fill_manual(values = colors) + -->
<!--   labs(x = "Response", y = "Error", title = "rRMSE", fill="Method") + -->
<!--   geom_hline(yintercept=1, linetype="dotted", size=1) + -->
<!--   theme_cowplot(font_size = 20) + -->
<!--   theme(plot.title = element_text(hjust = 0.5)) -->

<!-- print(p_rrmse_noout) -->
<!-- ``` -->

<!-- Here, we look at the elapsed time ($log_2$ seconds) of each method. Note that the *mr.mash* run time does not include the run time of group-LASSO (but should be considered since we used it to initialize *mr.mash*). -->

<!-- ```{r run time low PVE, fig.height=12, fig.width=15, message=FALSE} -->
<!-- dsc_plots_time <- dsc_plots[which(dsc_plots$response==1 & dsc_plots$score_metric=="scaled_rrmse"),  -->
<!--                           -which(colnames(dsc_plots) %in% c("score_metric", "score_value", "response"))] -->

<!-- p_time <- ggplot(dsc_plots_time, aes_string(x = "method_fac", y = "time", fill = "method_fac")) + -->
<!--   geom_boxplot(color = "black", outlier.size = 1, width = 0.85) + -->
<!--   scale_fill_manual(values = colors) + -->
<!--   scale_y_continuous(trans="log2", breaks = trans_breaks("log2", function(x) 2^x), -->
<!--                      labels = trans_format("log2", math_format(2^.x))) + -->
<!--   labs(x = "", y = "Elapsed time (seconds) in log2 scale", title = "Run time", fill="Method") + -->
<!--   theme_cowplot(font_size = 20) + -->
<!--   theme(axis.line.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(), -->
<!--         plot.title = element_text(hjust = 0.5)) -->

<!-- print(p_time) -->
<!-- ``` -->


# Different PVE

```{r load data diff pve, warning=FALSE, message=FALSE}
###Load the dsc results
dsc_out <- dscquery("output/mvreg_all_genes_prior_GTExrealX_indepV_sharedB_2blocksr10_diffPVE", 
                    c("simulate.p_causal", "simulate.r", "simulate.r_causal", "simulate.pve", "simulate.B_cor", "simulate.w", 
                      "simulate.B_scale", "simulate.V_cor", "simulate", "fit", "score", "score.err", "fit.time"), 
                    groups="fit: mr_mash_em_can, mr_mash_em_data, mr_mash_em_dataAndcan, mr_mash_em_dataAndcan_dropcomp, mlasso, mtlasso, enet",
                    verbose=FALSE)

###Obtain simulation parameters
prop_testset <- 0.2
p_causal <- unique(dsc_out$simulate.p_causal)
r <- unique(dsc_out$simulate.r)
r_causal <- eval(parse(text=unique(dsc_out$simulate.r_causal)))
pve <- eval(parse(text=unique(dsc_out$simulate.pve)))
w <- eval(parse(text=unique(dsc_out$simulate.w)))
B_cor <- eval(parse(text=unique(dsc_out$simulate.B_cor)))
B_scale <- eval(parse(text=unique(dsc_out$simulate.B_scale)))
V_cor <- unique(dsc_out$simulate.V_cor)

###Remove list elements that are not useful anymore
dsc_out$simulate.r_causal <- NULL
dsc_out$simulate.w <- NULL
dsc_out$simulate.B_cor <- NULL
dsc_out$simulate.B_scale <- NULL
dsc_out$simulate.V_cor <- NULL
```

In this section the set up is exactly the same but with the first three responses having PVE=0.2 and the remaining seven responses having PVE=0.05.

## mr.mash vs other methods

```{r accuracy vs others diff PVE, fig.height=12, fig.width=15, message=FALSE}
###Convert from list to data.frame for plotting
dsc_plots <- convert_dsc_to_dataframe_gtex(dsc_out)

###Compute rmse score (relative to mr_mash_consec_em) and add it to the data
rrmse_dat <- compute_rrmse(dsc_plots)
dsc_plots <- rbind(dsc_plots, rrmse_dat)

###Remove mse from scores and keep only methods wanted
dsc_plots <- dsc_plots[which(dsc_plots$score_metric!="scaled_rmse" ), ]

###Create factor version of method
dsc_plots$method_fac <- factor(dsc_plots$method, levels=c("mr_mash_em_can", "mr_mash_em_data",
                                                          "mr_mash_em_dataAndcan", "mr_mash_em_dataAndcan_dropcomp", 
                                                          "mlasso", "mtlasso", "enet"),
                                                labels=c("mr_mash_can", "mr_mash_data", "mr_mash_both", 
                                                         "mr_mash_both_drop", "mlasso", "mtlasso", "enet"))

###Create factor version of response
dsc_plots$response_fac <- as.factor(dsc_plots$response)

###Create plots
p_rrmse <- ggplot(dsc_plots[which(dsc_plots$score_metric=="scaled_rrmse"), ], aes_string(x = "response_fac", y = "score_value", fill = "method_fac")) +
  geom_boxplot(color = "black", outlier.size = 1, width = 0.85) +
  scale_fill_manual(values = colors) +
  labs(x = "Response", y = "Error", title = "rRMSE", fill="Method") +
  geom_hline(yintercept=1, linetype="dotted", size=1) +
  theme_cowplot(font_size = 20) +
  theme(plot.title = element_text(hjust = 0.5))

print(p_rrmse)
```

Let's now remove outliers from the plots to make things a little clearer.

```{r accuracy no outliers diff PVE, fig.height=12, fig.width=15, message=FALSE}
p_rrmse_noout <- ggplot(dsc_plots[which(dsc_plots$score_metric=="scaled_rrmse"), ], aes_string(x = "response_fac", y = "score_value", fill = "method_fac")) +
  Ipaper::geom_boxplot2(color = "black",width.errorbar = 0, width = 0.85) +
  scale_fill_manual(values = colors) +
  labs(x = "Response", y = "Error", title = "rRMSE", fill="Method") +
  geom_hline(yintercept=1, linetype="dotted", size=1) +
  theme_cowplot(font_size = 20) +
  theme(plot.title = element_text(hjust = 0.5))

print(p_rrmse_noout)
```

Here, we look at the elapsed time ($log_2$ seconds) of each method. Note that the *mr.mash* run time does not include the run time of group-LASSO (but should be considered since we used it to initialize *mr.mash*).

```{r run time diff PVE, fig.height=12, fig.width=15, message=FALSE}
dsc_plots_time <- dsc_plots[which(dsc_plots$response==1 & dsc_plots$score_metric=="scaled_rrmse"), 
                          -which(colnames(dsc_plots) %in% c("score_metric", "score_value", "response"))]

p_time <- ggplot(dsc_plots_time, aes_string(x = "method_fac", y = "time", fill = "method_fac")) +
  geom_boxplot(color = "black", outlier.size = 1, width = 0.85) +
  scale_fill_manual(values = colors) +
  scale_y_continuous(trans="log2", breaks = trans_breaks("log2", function(x) 2^x),
                     labels = trans_format("log2", math_format(2^.x))) +
  labs(x = "", y = "Elapsed time (seconds) in log2 scale", title = "Run time", fill="Method") +
  theme_cowplot(font_size = 20) +
  theme(axis.line.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
        plot.title = element_text(hjust = 0.5))

print(p_time)
```

